stack-deploy/src/instance_spec.rs

```diff
@@ -8,6 +8,42 @@ pub(crate) struct RemoteOperation {
 
 pub type Capabilities = std::collections::BTreeSet<aws_sdk_cloudformation::types::Capability>;
 
+pub struct TemplateUploader<'a> {
+    cloudformation: &'a aws_sdk_cloudformation::Client,
+    s3: &'a aws_sdk_s3::Client,
+    s3_bucket_name_output_key: &'a OutputKey,
+    stack_name: &'a StackName,
+}
+
+impl TemplateUploader<'_> {
+    async fn upload(&self, template_rendered: &TemplateRendered) -> TemplateUrl {
+        let hex =
+            hex::encode(<sha2::Sha256 as sha2::Digest>::digest(&template_rendered.body).as_slice());
+
+        let s3_bucket_name = crate::stack::read_stack_output(
+            self.cloudformation,
+            self.stack_name,
+            self.s3_bucket_name_output_key,
+        )
+        .await;
+
+        let s3_object_key = format!("{hex}.{}", template_rendered.format.file_ext());
+
+        log::info!("Uploading template to: {s3_bucket_name}/{s3_object_key}");
+
+        self.s3
+            .put_object()
+            .bucket(s3_bucket_name)
+            .key(s3_object_key)
+            .body((&template_rendered.body).into())
+            .send()
+            .await
+            .unwrap();
+
+        "https://s3.amazonaws.com/{s3_bucket_name}/{s3_object_key}".into()
```

Comment:
__question:__ doesn't this return a string with the curly braces not interpolated? Should this use `format!(...)`?